pkg/providers/ovirt/mappings/merger.go

```diff
@@ -0,0 +1,83 @@
+package mappings
+
+import (
+	v2vv1alpha1 "github.com/kubevirt/vm-import-operator/pkg/apis/v2v/v1alpha1"
+	"github.com/kubevirt/vm-import-operator/pkg/utils"
+)
+
+// MergeMappings creates new resource mapping spec containing all of the mappings from the externalMappingSpec with all mappings from crMappings. Mappings from the crMappings will overwrite ones from the external mapping.
+func MergeMappings(externalMappingSpec *v2vv1alpha1.ResourceMappingSpec, vmiMapping *v2vv1alpha1.OvirtMappings) v2vv1alpha1.OvirtMappings {
+	if externalMappingSpec == nil && vmiMapping == nil {
+		return v2vv1alpha1.OvirtMappings{}
+	}
+	primaryMappings, secondaryMappings := extractMappings(externalMappingSpec, vmiMapping)
+
+	networkMappings := mergeMappings(primaryMappings.NetworkMappings, secondaryMappings.NetworkMappings)
+	storageMappings := mergeMappings(primaryMappings.StorageMappings, secondaryMappings.StorageMappings)
+
+	ovirtMappings := v2vv1alpha1.OvirtMappings{
+		NetworkMappings: networkMappings,
+		StorageMappings: storageMappings,
+	}
+	return ovirtMappings
+}
+
+func extractMappings(externalMappingSpec *v2vv1alpha1.ResourceMappingSpec, crMappings *v2vv1alpha1.OvirtMappings) (primaryMappings *v2vv1alpha1.OvirtMappings, secondaryMappings *v2vv1alpha1.OvirtMappings) {
+	if crMappings == nil {
+		primaryMappings = &v2vv1alpha1.OvirtMappings{}
+	} else {
+		primaryMappings = crMappings
+	}
+
+	if externalMappingSpec == nil {
+		secondaryMappings = &v2vv1alpha1.OvirtMappings{}
+	} else {
+		secondaryMappings = externalMappingSpec.OvirtMappings
+	}
+	return
+}
+func mergeMappings(primaryMappings *[]v2vv1alpha1.ResourceMappingItem, secondaryMappings *[]v2vv1alpha1.ResourceMappingItem) *[]v2vv1alpha1.ResourceMappingItem {
+	var mapping []v2vv1alpha1.ResourceMappingItem
+
+	if primaryMappings == nil {
+		return secondaryMappings
+	}
+	if secondaryMappings == nil {
+		return primaryMappings
+	}
+	secondaryIDMap, secondaryNameMap := utils.IndexByIDAndName(secondaryMappings)
+	usedIDs := make(map[string]bool)
+	// Copy everything from the primary mapping to the output
+	for _, item := range *primaryMappings {
+		id := item.Source.ID
+		name := item.Source.Name
+		if id == nil && name == nil {
+			break
+		}
+		mapping = append(mapping, item)
+		// Delete from the secondary what we've already used
+		if id != nil {
+			usedIDs[*id] = true
+			delete(secondaryIDMap, *id)
+		}
```

Comment:
What about using else if here? If we have both name and id we will attempt to delete it twice.

Reply:
In the other if's body data is removed from the "by-name" secondary mapping index. Will add test that shows the use-case for two separate ifs.